components/support/nimbus-fml/src/structure/validator.rs

```diff
@@ -0,0 +1,485 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+* License, v. 2.0. If a copy of the MPL was not distributed with this
+* file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::error::FMLError;
+use crate::intermediate_representation::{FeatureDef, TypeFinder, TypeRef};
+use crate::{
+    error::Result,
+    intermediate_representation::{EnumDef, ObjectDef},
+};
+use std::collections::{BTreeMap, HashSet};
+
+pub(crate) struct StructureValidator<'a> {
+    enum_defs: &'a BTreeMap<String, EnumDef>,
+    object_defs: &'a BTreeMap<String, ObjectDef>,
+}
+
+impl<'a> StructureValidator<'a> {
+    pub(crate) fn new(
+        enums: &'a BTreeMap<String, EnumDef>,
+        objs: &'a BTreeMap<String, ObjectDef>,
+    ) -> Self {
+        Self {
+            enum_defs: enums,
+            object_defs: objs,
+        }
+    }
+
+    fn _get_enum(&self, nm: &str) -> Option<&EnumDef> {
+        self.enum_defs.get(nm)
+    }
+
+    fn get_object(&self, nm: &str) -> Option<&ObjectDef> {
+        self.object_defs.get(nm)
+    }
+
+    pub(crate) fn validate_object_def(&self, object_def: &ObjectDef) -> Result<()> {
+        let obj_nm = &object_def.name;
+        for prop in &object_def.props {
+            let prop_nm = &prop.name;
+
+            // Check the types exist for this property.
+            let path = format!("objects/{obj_nm}/{prop_nm}");
+            self.validate_type_ref(&path, &prop.typ)?;
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn validate_feature_def(&self, feature_def: &FeatureDef) -> Result<()> {
```

Comment:
This method is the largest non-mechanical change: here we're aggregating a lot of disparate methods which were all iterating over `feature_def.props`, and put them all in one loop.